src/ai/backend/common/events/hub/hub.py

```diff
@@ -0,0 +1,176 @@
+        del self._propagators[propagator_id]
+
+    async def propagate_event(self, event: AbstractEvent) -> None:
+        """
+        Propagate an event to all registered propagators.
+        :param event: The event to be propagated.
+        """
+        domain_id = event.domain_id()
+        if domain_id is None:
+            # If the event does not have a domain ID, it is not propagated.
+            return
+        propagators = self._get_propagators_by_alias(event.event_domain(), domain_id)
+        for propagator in propagators:
+            await propagator.propagate_event(event)
+
+    async def close_by_alias(
+        self,
+        alias_domain: EventDomain,
+        alias_id: str,
+    ) -> None:
+        """
+        Close all propagators associated with the specified alias.
+        :param alias_domain: The domain of the alias (e.g., USER, SESSION, KERNEL).
+        :param alias_id: The ID of the alias.
+        """
+        if (alias_domain, alias_id) not in self._key_alias:
+            raise ValueError(f"Alias {alias_domain}:{alias_id} not found.")
+
+        propagator_set = self._key_alias[(alias_domain, alias_id)]
+        for propagator_id in propagator_set:
+            await self._propagators[propagator_id].propagator.close()
+
+    def _get_propagators_by_alias(
+        self,
+        alias_domain: EventDomain,
+        alias_id: str,
+    ) -> set[EventPropagator]:
+        """
+        Get the propagator associated with the specified alias.
+        :param alias_domain: The domain of the alias (e.g., USER, SESSION, KERNEL).
+        :param alias_id: The ID of the alias.
+        :return: The propagator associated with the alias, or None if not found.
+        """
+        if (alias_domain, alias_id) not in self._key_alias:
+            return set()
+        propagator_set = self._key_alias[(alias_domain, alias_id)]
+        propagators = set()
+        for propagator_id in propagator_set.copy():
+            if propagator_id in self._propagators:
+                propagators.add(self._propagators[propagator_id].propagator)
+        return propagators
+
+    def _add_alias(
+        self,
+        alias_key: tuple[EventDomain, str],
+        propagator_id: uuid.UUID,
+    ) -> None:
+        """
+        Add an alias for a propagator.
+        :param alias_domain: The domain of the alias (e.g., USER, SESSION, KERNEL).
+        :param alias_id: The ID of the alias.
+        :param propagator_id: Unique identifier for the propagator.
+        """
+        if alias_key not in self._key_alias:
+            self._key_alias[alias_key] = set()
+        self._key_alias[alias_key].add(propagator_id)
+
+    def _remove_alias(
+        self,
+        alias_key: tuple[EventDomain, str],
+        propagator_id: uuid.UUID,
+    ) -> None:
+        """
+        Remove an alias for a propagator.
+        :param alias_domain: The domain of the alias (e.g., USER, SESSION, KERNEL).
+        :param alias_id: The ID of the alias.
```

Comment:
This should be alias_key

Reply:
I changed the implementation direction once, so the comments are a bit broken.